compiler/src/yul/abi/functions.rs

```diff
@@ -0,0 +1,205 @@
+                },
+        } => (
+            literal_expression! { (data_size) },
+            literal_expression! { (padded_size) },
+        ),
+        AbiType::Array { .. } => unimplemented!(),
+    };
+
+    block_statement! {
+        (let array_size := [array_size])
+        (let array_data_size := mul(array_size, [inner_padded_size.clone()]))
+        (for {(let i := 0)} (lt(i, array_size)) {(i := add(i, 1))}
+        {
+            (let val_ptr := add([val], (mul(i, [inner_data_size.clone()]))))
+            (let val := mloadn(val_ptr, [inner_data_size]))
+            (pop((alloc_mstoren(val, [inner_padded_size]))))
+        })
+        (pop((alloc((sub((ceil32(array_data_size)), array_data_size))))))
+    }
+}
+
+pub fn decode<T: AbiEncoding>(typ: T, loc: AbiDecodeLocation) -> yul::Statement {
+    let func_name = decode_name(&typ, loc.clone());
+
+    let decode_expr = match typ.abi_type() {
+        AbiType::Uint { .. } => decode_uint(loc),
+        AbiType::Array { inner, size } => decode_array(*inner, size, loc),
+    };
+
+    return function_definition! {
+         function [func_name](start, head_ptr) -> val {
+            (val := [decode_expr])
+         }
+    };
+}
+
+fn decode_uint(loc: AbiDecodeLocation) -> yul::Expression {
+    match loc {
+        AbiDecodeLocation::Memory => expression! { mload(head_ptr) },
+        AbiDecodeLocation::Calldata => expression! { calldataload(head_ptr) },
+    }
+}
+
+fn decode_array(inner: AbiType, size: AbiArraySize, loc: AbiDecodeLocation) -> yul::Expression {
+    match inner {
+        AbiType::Array { .. } => unimplemented!(),
+        AbiType::Uint {
+            size:
+                AbiUintSize {
+                    padded_size,
+                    data_size,
+                },
+        } if padded_size == data_size => match size {
+            AbiArraySize::Static { size } => {
+                let total_size = size * data_size;
+                let total_size = literal_expression! { (total_size) };
+
+                match loc {
+                    AbiDecodeLocation::Calldata => expression! { ccopy(head_ptr, [total_size]) },
+                    AbiDecodeLocation::Memory => expression! { head_ptr },
+                }
+            }
+            AbiArraySize::Dynamic => match loc {
+                AbiDecodeLocation::Calldata => {
+                    let data_ptr = expression! { add(start, (calldataload(head_ptr))) };
+                    let array_size = expression! { calldataload([data_ptr.clone()]) };
+                    let data_size = literal_expression! { (data_size) };
+                    let total_size = expression! { add(32, (mul([array_size], [data_size]))) };
+
+                    expression! { ccopy([data_ptr], [total_size]) }
+                }
+                AbiDecodeLocation::Memory => {
+                    expression! { add(start, (mload(head_ptr))) }
+                }
+            },
+        },
+        AbiType::Uint { .. } => unimplemented!(),
```

Comment:
the above arm on line 175 matches anything where `padded_size == data_size`, which means that no packing needs to take place. the arm tagged here is intended to handle anything that needs to be packed

for the time being, we are unable to decode arrays that contain things like addresses, since they are padded in the abi encoding. once pack/unpack operations are implemented, we can support this